### PR TITLE
Tracing: http semconv update

### LINKED ISSUE #39430: Tracing: Adopt OpenTelemetry `error.type` attribute for logical calls
### Library name

Azure.Core

### Please describe the feature.

OTel defined `error.type` attribute we can use to report low-cardinality failures with. 

We can just add it in https://github.com/Azure/azure-sdk-for-net/blob/2b941a2f3cb6ebefba2477effc741941f1883ac8/sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs#L143

```csharp
public void Failed(Exception? exception = default)
{
    if (exception != null)
    {
        AddAttribute("error.type", exception.GetType().FullName);
    }

    _activityAdapter?.MarkFailed(exception);
}
```

(as long as it does not conflict with HTTP-level `error.type` we're going to set in https://github.com/Azure/azure-sdk-for-net/issues/39429)

### PR DESCRIPTION
**Pull request by lmolkova**
Fix #39430 and #39429

### CODE REVIEW

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs [hunk 1]
```diff
@@ -18,7 +19,7 @@ namespace Azure.Core.Pipeline
     {
         private const string AzureSdkScopeLabel = "az.sdk.scope";
         internal const string OpenTelemetrySchemaAttribute = "az.schema_url";
-        internal const string OpenTelemetrySchemaVersion = "https://opentelemetry.io/schemas/1.17.0";
+        internal const string OpenTelemetrySchemaVersion = "https://opentelemetry.io/schemas/1.23.0";
```
Comment (lmolkova) [inline sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:22]: should be released by the end of the week

File: sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs [hunk 2]
```diff
@@ -62,32 +62,34 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
...
             if (isActivitySourceEnabled && message.Request.Uri.Host is string host)
             {
-                scope.AddAttribute("net.peer.name", host);
-                int port = message.Request.Uri.Port;
-                if (port != 443)
-                {
-                    scope.AddIntegerAttribute("net.peer.port", port);
-                }
+                scope.AddAttribute("server.address", host);
+                scope.AddIntegerAttribute("server.port", message.Request.Uri.Port);
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs:78]: The requirement to not add for default port was removed?

File: sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs [hunk 3]
```diff
@@ -62,32 +62,34 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
...
             }
 
             if (_resourceProviderNamespace != null)
             {
                 scope.AddAttribute("az.namespace", _resourceProviderNamespace);
             }
 
-            if (message.Request.Headers.TryGetValue("User-Agent", out string? userAgent))
+            if (!isActivitySourceEnabled && message.Request.Headers.TryGetValue("User-Agent", out string? userAgent))
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs:86]: user_agent should not be added for OTel?

File: sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs [hunk 4]
```diff
@@ -62,32 +62,34 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
...
             {
                 scope.AddAttribute("http.user_agent", userAgent);
             }
 
             scope.Start();
+            if (scope.IsEnabled)
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs:92]: Seems to be no if block?

File: sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs [hunk 5]
```diff
@@ -106,13 +108,14 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
                 throw;
             }
 
+            string statusCodeStr = message.Response.Status.ToString(CultureInfo.InvariantCulture);
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs:111]: Why cache it if we don't use it in both branches?

File: sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs [hunk 6]
```diff
@@ -123,10 +126,17 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
...
             {
-                scope.AddAttribute("otel.status_code", "ERROR");
-                scope.Failed();
+                if (isActivitySourceEnabled)
+                {
+                    scope.AddAttribute("error.type", statusCodeStr);
+                }
+                else
+                {
+                    scope.AddAttribute("otel.status_code", "ERROR");
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs:135]: A bit confused why this only applies to diagnostic source

File: sdk/core/Azure.Core/tests/ClientDiagnosticsTests.cs [hunk 7]
```diff
@@ -273,6 +273,25 @@ public void FailedStopsActivityAndWritesExceptionEvent()
             CollectionAssert.Contains(activity.Tags, new KeyValuePair<string, string>("az.namespace", "Microsoft.Azure.Core.Cool.Tests"));
         }
 
+        [Test]
+        public void FailedStopsActivityAndWritesErrorType()
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/tests/ClientDiagnosticsTests.cs:277]: Should this go in the ClientDiagnosticsTests.Net50 file?

File: sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs (hunk 2 above)
Comment (lmolkova) [reply sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs:78]: will be removed very soon - https://github.com/open-telemetry/semantic-conventions/pull/459

File: sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs (hunk 3 above)
Comment (lmolkova) [reply sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs:86]: yep, gone - https://github.com/open-telemetry/semantic-conventions/pull/468

File: sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs (hunk 6 above)
Comment (lmolkova) [reply sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs:135]: `otel.status_code` is a back-compat thing

File: sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs (hunk 4 above)
Comment (lmolkova) [reply sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs:92]: oops

File: sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs (hunk 5 above)
Comment (lmolkova) [reply sdk/core/Azure.Core/src/Pipeline/Internal/RequestActivityPolicy.cs:111]: used on line 137 as well (effectively on both branches)

**Review by JoshLove-msft — APPROVED** [review]

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs (hunk 1 above)
Comment (lmolkova) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:22]: released :)

**Pull request merged**